### ISSUE TITLE
Use Meilisearch-rust for our integration-test

### ISSUE THREAD

**Issue by irevoire** [meilisearch/meilisearch#3008]
After rewriting a big chunk of our codebase and integration tests I noticed we were spending a large amount of time rewriting exactly the same functionality already available in [meilisearch-rust](https://github.com/meilisearch/meilisearch-rust).

PRO:
- From what I see, absolutely all the functions we uses in our integration tests exists in Meilisearch-rust and works better. 
- We have no tests on our integration test which means, sometimes we spend time trying to debug in meilisearch while its due to the complexity of our integration test / wrapper.
- Usually Meilisearch-rust gives us a better and more versatile API.
- If we're missing a feature one day for our test and we implements it, it'll benefit all the users of the integration as-well.
- When we create a new feature, we're going to automatically update the sdk while implementing it.
- While writing new features we should realise faster if the API is badly designed since we're going to implements it ourselves directly.

CONS:
- When creating new features we'll need to write the documentation as well as the code (while currently everything is undocumented and thus faster to write).
- If we want to add functions to the sdk to facilitate our life, what's the policy? We'll need to work closely with the @meilisearch/integration-team and this may slow us down.
- If the @meilisearch/integration-team needs to do a breaking change on Meilisearch-rust it'll break our test suite as well (or at least we'll need to spend time updating our test suite)

Here is a random picked example of a complex route duplicated between Meilisearch-rust and meilisearch:
- meilisearch-rust
```rust
    pub async fn wait_for_task(
        &self,
        task_id: impl AsRef<u32>,
        interval: Option<Duration>,
        timeout: Option<Duration>,
    ) -> Result<Task, Error> {
        let interval = interval.unwrap_or_else(|| Duration::from_millis(50));
        let timeout = timeout.unwrap_or_else(|| Duration::from_millis(5000));

        let mut elapsed_time = Duration::new(0, 0);
        let mut task_result: Result<Task, Error>;

        while timeout > elapsed_time {
            task_result = self.get_task(&task_id).await;
            match task_result {
                Ok(status) => match status {
                    Task::Failed { .. } | Task::Succeeded { .. } => {
                        return self.get_task(task_id).await;
                    }
                    Task::Enqueued { .. } | Task::Processing { .. } => {
                        elapsed_time += interval;
                        async_sleep(interval).await;
                    }
                },
                Err(error) => return Err(error),
            };
        }

        Err(Error::Timeout)
    }
```

- meilisearch
```rust
    pub async fn wait_task(&self, update_id: u64) -> Value {
        // try several times to get status, or panic to not wait forever
        let url = format!("/tasks/{}", update_id);
        for _ in 0..100 {
            let (response, status_code) = self.service.get(&url).await;
            assert_eq!(200, status_code, "response: {}", response);

            if response["status"] == "succeeded" || response["status"] == "failed" {
                return response;
            }

            // wait 0.5 second.
            sleep(Duration::from_millis(500)).await;
        }
        panic!("Timeout waiting for update id");
    }
```

**Comment by brunoocasali** [comment]
Hi @irevoire, thanks for raising this with us!

I like the idea, this approach will create a bonding between the teams, and I truly believe it could only have benefits for both sides.

> If we want to add functions to the sdk to facilitate our life, what's the policy? We'll need to work closely with the https://github.com/orgs/meilisearch/teams/integration-team and this may slow us down.

What kind of functions are you talking about? Public accessible functions or just internal and structural code?
In the integrations team, we try, whenever possible, to have a standard style between our SDKs.

> If the https://github.com/orgs/meilisearch/teams/integration-team needs to do a breaking change on Meilisearch-rust it'll break our test suite as well (and we will need to spend time updating our test suite because to write a new feature we need to update Meilisearch-rust to the latest version)

I didn't get this point, you mean you'll make "meilisearch-rust" a dependency of "milli" or "meilisearch"? Otherwise, I can't think of any breaking change that will hurt the test suite.

Some points I would like to clarify:

- The responsibilities of the engine team would be:
  - Implement the new features when a Meilisearch release is out (pre-release weeks).
  - Do code reviews.
  - Maintenance in general (bug fixes and such).
  - Will we have a focal point in the engine team when the integrations team has questions?

- The responsibilities of the integrations team would be:
  - Help to shape the public API of the SDK.
  - Make sure the SDK is helpful for the users.
  - Do code reviews.
  - Do the triage when a new issue comes.
  - Maintenance in general (bug fixes and such).
 
 
 Let me know what you think about it...
 
 📡 to @meilisearch/integration-team

**Comment by irevoire** [comment]
> What kind of functions are you talking about? Public accessible functions or just internal and structural code?
In the integrations team, we try, whenever possible, to have a standard style between our SDKs.

Both internal ones are not important, I guess.

But for public functions, we might want to add some new features that simplify our life.
For example, off the top of my head, I could easily see us implementing a `next` method on the paginated routes :thinking: 
Maybe even implements the official rust iterator on top of paginated resources.
Or simplifying the parameters we push in the `task::wait_for_completion` method because writing `None, None` everywhere is quite cumbersome :grimacing: 

We might want to update the task type that's quite complex currently for a simpler one as well; I think :thinking: 
And probably a lot of other little details here and there...


> I didn't get this point; you mean you'll make "meilisearch-rust" a dependency of "milli" or "meilisearch"? Otherwise, I can't think of any breaking change that will hurt the test suite.

For example, if one day, the integration team decide that all the SDK should rename the `key` module to `keys`, then it would break all our keys-related tests :grimacing: 
I don't think it's a big issue, personally, but it'll probably happen one day since you try to keep all the SDKs quite similar.

-------

> Let me know what you think about it...

Your list of bullets points looks perfect to me :ok_hand:

**Comment by bidoubiwa** [comment]
> For example, if one day, the integration team decide that all the SDK should rename the key module to keys, then it would break all our keys-related tests 😬
> I don't think it's a big issue, personally, but it'll probably happen one day since you try to keep all the SDKs quite similar.

After the release of a v1 version, we should not have these kind of breaking. If we do, it will require a v2 release and a sync with the engine team to ensure your tests are not failing, and/or implement a backward compatibility.

**Comment by irevoire** [comment]
Ok, so that means we won't be able to break the SDK ourselves either without doing a v2.0 of meilisearch?

**Comment by brunoocasali** [comment]
> Ok, so that means we won't be able to break the SDK ourselves either without doing a v2.0 of meilisearch?

You definitely can, because the meilisearch-rust is still beta.

**Comment by irevoire** [comment]
Nice!

**Comment by brunoocasali** [comment]
I don't think it is a big deal if you want to change how the public API works because we will help to review the code in the end, so if we disagree with some decision, we can discuss it.

Also, the engine team has expertise in the rust stack, so we trust them to make those decisions.

By the way, regarding the idea of making a standard style between SDKs, I don't know if we will be able to support this for much longer because the meilisearch-rust is already very different than the others.

Personally, having an idiomatic API is the most important. The people from the Rust community should be comfortable using their Rust SDK, and the same applies to all the other SDKs. And I think keeping a standard styling between SDKs and being idiomatic are ideas that go in opposite directions (and is up to us the integrations team to keep that balance).
